platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -246,10 +251,16 @@ public void add(Class<?> serviceType) {
             }
 
             @Override
-            public void add(Class<?> serviceType, Class<?> implementationType) {
+            public <S, I extends S> void add(Class<S> serviceType, Class<I> implementationType) {
                 ownServices.add(new ConstructorService(DefaultServiceRegistry.this, serviceType, implementationType));
             }
 
+            @Override
+            public <S, I extends S> void add(Class<S> serviceType1, Class<?> serviceType2, Class<I> implementationType) {
```

Comment:
Maybe add some unit test coverage for this.

Reply:
[Done](https://github.com/gradle/gradle/pull/29560/files#diff-9b47100899a0e5e1b1f369873022517083cba003847b36b758199b56ae47d083)